bundles/org.openhab.binding.tsmart/src/main/java/org/openhab/binding/tsmart/internal/TSmartUDPListener.java

```
@@ -0,0 +1,143 @@
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.tsmart.internal;
+
+import static org.openhab.binding.tsmart.internal.TSmartBindingConstants.T_SMART_PORT;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.SocketTimeoutException;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link TSmartUDPListener} is responsible for listening for UDP packets and
+ * routing the information to the correct ThingHandler or Discovery Service
+ *
+ * @author James Melville - Initial contribution
+ */
+@NonNullByDefault
+public class TSmartUDPListener implements Runnable {
+
+    private final Logger logger = LoggerFactory.getLogger(TSmartUDPListener.class);
+
+    private static @Nullable TSmartUDPListener listener;
+    private static @Nullable TSmartDiscoveryService discoService;
+    private volatile static Map<InetAddress, TSmartHandler> handlerAddressMap = new ConcurrentHashMap<InetAddress, TSmartHandler>();
+
+    private static void enableListener() {
+        if (listener == null) {
+            listener = new TSmartUDPListener();
+            Thread t = new Thread(listener);
+            t.start();
+        }
+    }
+
+    /**
+     * Maintain a mapping between a device address and the openHAB handler
+     * in order to route any received UDP packets to the correct handler.
+     * This method adds the address and ThingHandler to the map.
+     *
+     * @param addr Network address of device
+     * @param handler Instance of ThingHander
+     */
+    public static void addHandler(InetAddress addr, TSmartHandler handler) {
+        if (handlerAddressMap.containsKey(addr)) {
+            // throw error
```

That would probably make sense to fail fast on duplicate subscriptions.